### PR TITLE
`Programming exercises`: Fix issues with binary files when importing and starting programming exercises

### LINKED ISSUE #6440: `Programming Exercise`: Importing of Swift programming exercises is not possible
### Describe the bug

Importing Swift programming exercise (maybe this also occurs to other languages) currently does not work as intended. When clicking on the import button, no error message is shown. However, in the web console a 500 error is shown (see screenshot below). 
The exercise is created in the database and the template repository is set up. But the creation of the solution and test repository as well as both build plans fails.

### To Reproduce

1. Try to import an exercise (I tried to import a Swift exercise)
2. Click on 'Import'
3. No success or error message is shown
4. This error is shown in the web console:
Screenshot 2023-04-11 at 22 16 38
5. The exercise is created in the database and the template repository is set up
6. However, the solution and test repository as well as both build plans are not created.

### Expected behavior

The exercise should be imported successfully, all repositories and build plans should be set up and configured.

### Screenshots

_No response_

### Which version of Artemis are you seeing the problem on?

6.1.2

### What browsers are you seeing the problem on?

Safari

### Additional context

Java exercises are not affected.

### Relevant log output

_No response_

### LINKED ISSUE #6890: everything goes kabooom if you add a binary in template/solution repository
### Describe the bug

it is impossible to add a binary to a repository on bitbucket and then import the exercise in an exam. We (EIST team) changed the exercise so we don't have a binary on bitbucket but I think you should investigate this as we lost a lot of time trying to understand what was wrong.

MRE
Exercise: https://artemis.ase.in.tum.de/course-management/265/programming-exercises/10839
Repository: https://bitbucket.ase.in.tum.de/projects/EIST23TESTMREBREAKS/repos/eist23testmrebreaks-exercise/browse

### To Reproduce

.

### Expected behavior

.

### Screenshots

image

### Which version of Artemis are you seeing the problem on?

6.3.3

### What browsers are you seeing the problem on?

Chrome

### Additional context

_No response_

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by krusche**
Fixes #6440 and fixes #6890

#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://ls1intum.github.io/Artemis/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/server/).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context
There are different issues when dealing with binary files when importing and starting programming exercises.

### Description
* Ignore binary files during replacement
* Ignore a JGit exception when reading the response of a git push operation

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student

1. Create a programming exercise 
2. Add at least one binary file (e.g., add PNGs into the repo) to all three repos (template, solution, tests)
3. Import the exercise within Artemis
4. Also import the exercise from file (after downloading a zip of the exercise from the details page)
5. Start the two new imported exercises and participate from student's perspective
6. Make sure everything is working properly

#### Exam Mode Testing
Same as above

#### Performance Review
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2
#### Exam Mode Test
- [ ] Test 1
- [ ] Test 2

### Test Coverage
no changes

### Screenshots
no changes

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/domain/UserGroup.java [hunk 1]
```diff
@@ -18,12 +19,36 @@ public class UserGroup {
...
+            UserGroupKey that = (UserGroupKey) o;
+
+            if (!Objects.equals(userId, that.userId)) {
+                return false;
+            }
+            return Objects.equals(group, that.group);
+        }
+
+        @Override
+        public int hashCode() {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/domain/UserGroup.java:48]: generated with IntelliJ

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseImportService.java [hunk 2]
```diff
@@ -95,18 +95,18 @@ public void importRepositories(final ProgrammingExercise templateExercise, final
...
-            AuxiliaryRepository newAuxiliaryRepository = newExercise.getAuxiliaryRepositories().get(i);
-            newAuxiliaryRepository.setRepositoryUrl(repositoryUrl);
-            auxiliaryRepositoryRepository.save(newAuxiliaryRepository);
+        List<AuxiliaryRepository> auxRepos = templateExercise.getAuxiliaryRepositories();
+        for (int i = 0; i < auxRepos.size(); i++) {
+            AuxiliaryRepository auxRepo = auxRepos.get(i);
+            var repoUrl = versionControl.copyRepository(sourceProjectKey, auxRepo.getRepositoryName(), sourceBranch, targetProjectKey, auxRepo.getName()).toString();
+            AuxiliaryRepository newAuxRepos = newExercise.getAuxiliaryRepositories().get(i);
+            newAuxRepos.setRepositoryUrl(repoUrl);
+            auxiliaryRepositoryRepository.save(newAuxRepos);
```
Comment (julian-christl) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseImportService.java:109]:
```suggestion
            AuxiliaryRepository newAuxRepo = newExercise.getAuxiliaryRepositories().get(i);
            newAuxRepo.setRepositoryUrl(repoUrl);
            auxiliaryRepositoryRepository.save(newAuxRepo);
```

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseImportService.java [hunk 3]
```diff
@@ -95,18 +95,18 @@ public void importRepositories(final ProgrammingExercise templateExercise, final
...
-            AuxiliaryRepository auxiliaryRepository = auxiliaryRepositories.get(i);
-            String repositoryUrl = versionControl
-                    .copyRepository(sourceProjectKey, auxiliaryRepository.getRepositoryName(), sourceBranch, targetProjectKey, auxiliaryRepository.getName()).toString();
-            AuxiliaryRepository newAuxiliaryRepository = newExercise.getAuxiliaryRepositories().get(i);
-            newAuxiliaryRepository.setRepositoryUrl(repositoryUrl);
-            auxiliaryRepositoryRepository.save(newAuxiliaryRepository);
+        List<AuxiliaryRepository> auxRepos = templateExercise.getAuxiliaryRepositories();
+        for (int i = 0; i < auxRepos.size(); i++) {
+            AuxiliaryRepository auxRepo = auxRepos.get(i);
+            var repoUrl = versionControl.copyRepository(sourceProjectKey, auxRepo.getRepositoryName(), sourceBranch, targetProjectKey, auxRepo.getName()).toString();
```
Comment (julian-christl) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseImportService.java:106]:
```suggestion
            String repoUrl = versionControl.copyRepository(sourceProjectKey, auxRepo.getRepositoryName(), sourceBranch, targetProjectKey, auxRepo.getName()).toString();
```
Oder gibt es einen Grund `String` zu `var` zu verkürzen?

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java [hunk 4]
```diff
@@ -912,32 +910,47 @@ public void replaceVariablesInFileRecursive(String startPath, Map<String, String
...
+            // continue
         }
-        Files.writeString(replaceFilePath, fileContent, charset);
+    }
+
+    private static boolean isBinaryFile(Path filePath) {
+        // TODO: extend the list of potential binary files
+        var fileName = filePath.toString().toLowerCase();
+
+        return fileName.endsWith(".png") || fileName.endsWith(".jpg") || fileName.endsWith(".heic");
```
Comment (julian-christl) [inline src/main/java/de/tum/in/www1/artemis/service/FileService.java:940]: I'd suggest retrieving the filename from the path and then the extension using FileUtils.
Then, we can easily compare it to a static List that we can define above.

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java [hunk 5]
```diff
@@ -912,32 +910,47 @@ public void replaceVariablesInFileRecursive(String startPath, Map<String, String
...
-        Charset charset = StandardCharsets.UTF_8;
 
-        String fileContent = Files.readString(replaceFilePath, charset);
-        for (Map.Entry<String, String> replacement : replacements.entrySet()) {
-            fileContent = fileContent.replace(replacement.getKey(), replacement.getValue());
+        try {
+            if (isBinaryFile(filePath)) {
+                // do not try to read binary files with 'readString'
+                return;
+            }
```
Comment (julian-christl) [inline src/main/java/de/tum/in/www1/artemis/service/FileService.java:922]: We can move this out of the try-catch and avoid a nesting layer

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java [hunk 6]
```diff
@@ -912,32 +910,47 @@ public void replaceVariablesInFileRecursive(String startPath, Map<String, String
...
+            // Note: we cannot check if this is a binary file or not. We try to read it, in case it fails, we only log this below, but continue
```
Comment (julian-christl) [inline src/main/java/de/tum/in/www1/artemis/service/FileService.java:923]: I don't understand this note. If we cannot check this, why do we do something that looks like a check?
Looking at the code, I'd say the note is wrong. Can you explain this? Maybe rephrase the note?

**Review by julian-christl — COMMENTED** [review]
I assume that the exception check is correct, it should be covered by manual testing.

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java (hunk 6 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/service/FileService.java:923]: I initially wanted to have a universal checks, but this seems to be very difficult like mentioned here: https://stackoverflow.com/questions/620993/determining-binary-text-file-type-in-java

So there is no 100% guarantee and we need to run into the exception.

Then I thought, we can at least implement a simply check for typical binary files such images, however this will never be exhaustive and only prevents the exception from happening too often.

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java (hunk 4 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/service/FileService.java:940]: this was just a quick implementation, do you have a concrete suggestion how to implement this? if yes, feel free to provide it, otherwise I would move this to a follow-up to deploy the bugfix asp

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseImportService.java (hunk 3 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseImportService.java:106]: My goal was that the code fits into one line in IntelliJ to avoid line breaks which - at least for me - make the code harder to read ;-)

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java (hunk 4 above)
Comment (maximiliansoelch) [reply src/main/java/de/tum/in/www1/artemis/service/FileService.java]: @julian-christl I added your suggestion in https://github.com/ls1intum/Artemis/pull/7178/commits/26b55d71482cbf3f2cff4f2c354427bd8d30ad33 
Is that what you envisioned?

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java [hunk 7]
```diff
@@ -59,6 +59,13 @@ public class FileService implements DisposableBean {
 
     private final ScheduledExecutorService executor = Executors.newScheduledThreadPool(Runtime.getRuntime().availableProcessors());
 
+    /**
+     * A list of common binary file extensions.
+     * Extensions must be lower-case without leading dots.
+     */
+    private static final Set<String> binaryFileExtensions = Set.of("png", "jpg", "jpeg", "heic", "gif", "tiff", "psd", "pdf", "doc", "docx", "xls", "xlsx", "ppt", "pptx", "pages",
+            "numbers", "key", "odt", "zip", "rar", "7z", "tar", "iso", "mdb", "sqlite", "exe");
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/FileService.java:67]: With this. the explicit ignore of "gradle-wrapper.jar" can get removed.
```suggestion
    private static final Set<String> binaryFileExtensions = Set.of("png", "jpg", "jpeg", "heic", "gif", "tiff", "psd", "pdf", "doc", "docx", "xls", "xlsx", "ppt", "pptx", "pages",
            "numbers", "key", "odt", "zip", "rar", "7z", "tar", "iso", "mdb", "sqlite", "exe", "jar");
```

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java (hunk 4 above)
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/service/FileService.java]: Exactly what I envisioned, thank you!

File: src/main/java/de/tum/in/www1/artemis/service/FileService.java (hunk 6 above)
Comment (julian-christl) [reply src/main/java/de/tum/in/www1/artemis/service/FileService.java]: Thanks for the explanation. The updated note explains it very well.

**Review by julian-christl — DISMISSED** [review]
Thanks for the changes, code looks good to me now.

**Review by laurenzfb — DISMISSED** [review]
Code LGTM! Also tested on TS1, worked fine as expected

**Review by terlan98 — COMMENTED** [review]
Tested locally with a Java exercise containing .png and .docx files in all 3 repos. The binary files were imported properly and everything worked fine from student's perspective 👍 

⚠️ 
However, the test.json file was not in the test repository when I imported the exercise **from file**. This caused the student and solution build to fail. If it's unrelated, please let me know and I'll approve.
Screenshot 2023-09-12 at 14 01 56

**Comment by maximiliansoelch** [comment]
> Tested locally with a Java exercise containing .png and .docx files in all 3 repos. The binary files were imported properly and everything worked fine from student's perspective 👍
> 
> ⚠️ However, the test.json file was not in the test repository when I imported the exercise **from file**. This caused the student and solution build to fail. If it's unrelated, please let me know and I'll approve. Screenshot 2023-09-12 at 14 01 56

@terlan98 Is the test.json file included in the zip file when you uncompress it locally?

**Comment by terlan98** [comment]
> @terlan98 Is the test.json file included in the zip file when you uncompress it locally?
It is included in the original exercise's zip file but not in the imported one. So it got lost in the importing process somehow.

**Review by terlan98 — APPROVED** [review]
Tested the aforementioned `test.json` issue again. If the user changes the short name of the exercise while importing, it somehow breaks the test.json file. This problem is unrelated to this PR, so I'll approve. I'll create a GitHub issue for it.

**Review by pal03377 — APPROVED** [review]
Tested on my test server (which uses local VC/CI). I added a jpg file + sometimes a jpg without an extension, and tried importing the exercise through Artemis and using the ZIP export, also participating as a student after that. I did not find any issues 👍

**Review by Strohgelaender — CHANGES_REQUESTED** [review]
Please add the ".jar" extension to the list of ignored file extensions, this was what lead to the issue in #6890.

**Review by tobias-lippert — APPROVED** [review]
Code LGTM

**Review by julian-christl — APPROVED** [review]
Code reapproved

**Review by laurenzfb — APPROVED** [review]

**Review by terlan98 — APPROVED** [review]

**Review by Strohgelaender — APPROVED** [review]
